### PR TITLE
ramips: wevo: set uImage name and status led

### PR DESCRIPTION
**Pull request by mans0n**
~~- set uImage name similar to that of stock FW, so that the stock bootloader will not complain.~~
~~- remove WAN port index.~~
~~- load WAN MAC address directly from the flash.~~
~~- provide label MAC address.~~
~~- use USB LED for boot/failsafe/upgrade status indication.~~
~~- replace gpio-keys-polled with gpio-keys.~~
~~- increase flash's SPI frequency to 80MHz.~~
~~- add mt76 led nodes to make WiFi LEDs work.~~
~~- drop unnecessary pinmux groups.~~

The stock firmware and bootloader only accept uImage with names that match
certain patterns. This patch enables OpenWrt installation from stock
firmware without having to reflash the bootloader or access the UART
console.

Installation via web interface:
1.  Flash **initramfs** image through the stock web interface.
2.  Boot into OpenWrt and perform sysupgrade with sysupgrade image.

And also make use of the USB LED for boot/failsafe/upgrade status
indication. After boot it returns to its original role.

### CODE REVIEW

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -395,7 +393,9 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan:1" "1:lan:2" "2:lan:3" "3:lan:4" "6@eth0"
 		;;
-	netgear,r6120)
+	netgear,r6120|\
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
 		ucidef_add_switch "switch0" \
 			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "6@eth0"
```
Comment (adrianschmutzler) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Won't this result in a collision for port 4 in luci?

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -510,7 +510,9 @@ ramips_setup_macs()
 	alfa-network,w502u|\
 	arcwireless,freestation5|\
 	cudy,wr1000|\
-	netgear,wnce2001)
+	netgear,wnce2001|\
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
 		wan_mac=$(mtd_get_mac_binary factory 0x2e)
```
Comment (adrianschmutzler) [inline target/linux/ramips/base-files/etc/board.d/02_network]: The looks strange to me, as &ethernet pulls the address from <&factory 0xe000>.

Can you provide a hexdump of the whole factory partition (for both devices if possible)?

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: I'm not sure I can post the whole partition here.
The MAC addresses of 11acnas device I have are xx:xx:xx:99:xx:yy,
and I searched all "99" occurrences in the factory partition. Then I got:
0x4: xx:xx:xx:99:xx:80 (WiFi 2.4GHz)
0x2e: xx:xx:xx:99:xx:89 (WAN and LABEL)
0x8004: xx:xx:xx:99:xx:84 (WiFi 5GHz)
0xe000: xx:xx:xx:99:xx:88 (LAN)

I don't have w2914ns-v2 device, so it needs a test from someone else. But AFAIK the two devices are almost same, so I think it will work for w2914ns-v2 too.

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 1 above)
Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Hmm it seemed okay when I tested...

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 2 above)
Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Okay, can you check 0x28 and 0xe006 on your device, too?

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: It's blank.
0x28: fd:9b:ff:ff:ff:ff
0xe006: ff:ff:ff:ff:ff:ff

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Okay, they really have a strange setup there, but what you did looks correct to me. Thank you for the additional research.

Would still be interested in the same information about w2914ns-v2 if someone can get it.

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 1 above)
Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Okay, I've also found other similar entries in 02_network.
However, let me just change my question: What is your reason to remove the ":5" for the wan port?

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: AFAIK port indices are only used for sorting ethernet ports in the same role.
https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/network.js#L541
For this model, there is only one WAN port, and the index number "5" does not affect in any way, for now.
Still it can be leaved if you think we can use port index in some other ways in the future.

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/02_network]: I see your point. Still, if the ":5" has no negative effect, removing it looks to me like removing information with no reason for it. So I'd keep it.
However, to be honest I'm aware that the subject isn't worth the big discussion we are having here. So I will also be able to sleep at night if you remove it :-)

**Comment by perillamint** [comment]
I did some experiments with WeVO W2914NS v2 firmware. I'll share some interesting things which I discovered.

First, WeVO uses the same header format which ipTIME(aka. TOTOLINK in other countries except S. Korea) products use. It checks the whole checksum of firmware file after 0x40 bytes, not only uImage.
Second, when I apply my mkiptimefw implementation on OpenWRT firmware, it refuses to boot.

**Comment by mans0n** [comment]
@perillamint The stock firmware uses standard uImage format. The stock web UI refuses sysupgrade image, but will accept initramfs image. IMO we don't need special mkwevofw, but just a proper installation guide.

BTW could you please test this patch for w2914ns-v2 if you have one? https://github.com/openwrt/openwrt/pull/2336#discussion_r314983097

**Comment by mans0n** [comment]
Rebased.

**Comment by perillamint** [comment]
I can confirm stock bootloader of 11AC NAS Router happily accepts that uImage. However, It seems I lost W2914NS V2 U-Boot stock image. :(

@mans0n

**Comment by mans0n** [comment]
@perillamint That's a shame... Thanks for your confirm on 11acnas, anyway. :)

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 2 above)
Comment (perillamint) [reply target/linux/ramips/base-files/etc/board.d/02_network]: OK, I'll grab my factory of W2914NS-V2 today and will post the result of the analysis. today.

Comment (perillamint) [reply target/linux/ramips/base-files/etc/board.d/02_network]: It is same as above result.

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Great. That's the answer I've been waiting for. Thanks! :D

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Okay, so this is settled. Since you have tested this, I would add this information to the commit message, like:
"
Both devices have an untypical MAC address configuration (both tested!):
0x4: xx:xx:xx:xx:xx:80 (WiFi 2.4GHz)
0x28: fd:9b:ff:ff:ff:ff (unused)
0x2e: xx:xx:xx:xx:xx:89 (WAN and label)
0x8004: xx:xx:xx:xx:xx:84 (WiFi 5GHz)
0xe000: xx:xx:xx:xx:xx:88 (LAN)
0xe006: ff:ff:ff:ff:ff:ff
"

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Okay, added.

**Comment by mans0n** [comment]
Rebased to add label MAC.

**Comment by mans0n** [comment]
I just realized that WiFi 2.4GHz LED is also broken (Why haven't I noticed that earlier?), so added another mt76 led node.

**Comment by mans0n** [comment]
After a thorough review, I found a few more things to improve. Sorry for the frequent changes.
The commit is now quite big, do you think it should be split?

**Comment by mans0n** [comment]
Fixed some typos.

File: target/linux/ramips/dts/mt7621_wevo_11acnas.dts [hunk 3]
```diff
@@ -6,10 +6,16 @@
 	compatible = "wevo,11acnas", "wevo,w2914ns-v2", "mediatek,mt7621-soc";
 	model = "WeVO 11AC NAS Router";
 
+	aliases {
+		led-boot = &led_usb;
+		led-failsafe = &led_usb;
+		led-upgrade = &led_usb;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_wevo_11acnas.dts]: The LED used here should not be mistakable. Using USB, WiFi, LAN LEDs etc. is not a good idea.

So, either remove this entirely or at least make this a separate commit, so the rest is not delayed by this discussion.

File: target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dts [hunk 4]
```diff
@@ -5,10 +5,16 @@
 / {
 	model = "WeVO W2914NS v2";
 
+	aliases {
+		led-boot = &led_usb;
+		led-failsafe = &led_usb;
+		led-upgrade = &led_usb;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dts]: Same as above.

File: target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi [hunk 5]
```diff
@@ -11,19 +11,20 @@
...
 	keys {
-		compatible = "gpio-keys-polled";
-		poll-interval = <20>;
+		compatible = "gpio-keys";
 
 		reset {
 			label = "reset";
 			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
 			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi]: I thought debounce-interval is used for gpio-keys-polled?

**Comment by adrianschmutzler** [comment]
> The commit is now quite big, do you think it should be split?

Yes, please. You do not need a separate commit for everything, but try to break it down into some logical units.

File: target/linux/ramips/dts/mt7621_wevo_11acnas.dts (hunk 3 above)
Comment (mans0n) [reply target/linux/ramips/dts/mt7621_wevo_11acnas.dts]: Hmm... There's no device using USB LED for status indication so you may be right, but I still think it will be very helpful to have a status LED, at least during boot. I'll separate it as another commit then.

File: target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi (hunk 5 above)
Comment (mans0n) [reply target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi]: gpio-keys also uses debounce-interval so it will be okay.
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/input/gpio-keys.txt#L25

**Comment by mans0n** [comment]
Split into three commits.

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 6]
```diff
@@ -427,6 +427,10 @@ wavlink,wl-wn575a3)
 	ucidef_set_led_rssi "wifi-med" "wifi-med" "$boardname:green:wifi-med" "wlan1" "50" "84"
 	ucidef_set_led_rssi "wifi-high" "wifi-high" "$boardname:green:wifi-high" "wlan1" "85" "100"
 	;;
+wevo,11acnas|\
+wevo,w2914ns-v2)
+	ucidef_set_led_usbport "usb" "USB" "$boardname:green:usb" "usb1-port1" "usb1-port2" "usb2-port1"
```
Comment (adrianschmutzler) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: This seems redundant/overlapping with the triggers defined for the USB led. Do you need this to get back control after the diag.sh script was running?

**Comment by adrianschmutzler** [comment]
@mans0n Okay, I think the split is reasonably chosen.

@blocktrron Maybe have a look at this for review. In particular, we would need a verdict whether the approach in patch 3 is acceptable or not.

File: target/linux/ramips/base-files/etc/board.d/01_leds (hunk 6 above)
Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Yes, exactly.

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Okay, so this already shows why using the USB led for diag is not a good idea, as it requires "hacks" like this one. However, I'm not a core-developer, so I will leave this for other reviewers to decide.

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: In my defense, there's already some devices with status led that goes back to the original role after boot. https://github.com/openwrt/openwrt/commit/0c5a53255307ab532b2f875db452803b7f297874 https://github.com/openwrt/openwrt/commit/a9360452f0130bb39a8e732334a174cc3f8991fd https://github.com/openwrt/openwrt/commit/9f0e2335764674d0324dd90adb3e185cc8befe5f

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Yes, and I would say it depends on the reviewer whether stuff like that comes in or not.

BTW, you provided an excellent choice of reference here, as each case is different:
0c5a532 uses LAN led without trigger in DTS
a936045 uses dts trigger for one led and 02_network for the other, but uses both for diag
9f0e233 uses signal strength indicator

Especially the last case has been accepted several times recently (particularly for the TP-Link CPExxx devices in ath79 target), as signal strenght indicators do not blink typically and thus have a reduced potential to confuse the user.
However, note that 9f0e233 uses the green indicator for diag AND signaling although the red one is completely unused. So this could also have slipped in ...

Nevertheless, I intentionally commented about this without a clear indication, as it depends on taste.

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: BTW "usbport" trigger does not blink, the LED just turns on when USB devices inserted. That's why I thought USB LED could also be a nice status LED.

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: No blinking on "activity"?

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Nope. https://www.kernel.org/doc/html/latest/leds/ledtrig-usbport.html
"disk-activity" might blink, but "usbport" does not have any blink option AFAIK.

Comment (adrianschmutzler) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Okay, I'd say that this raises your chances :-)

If used as diag LED, one might consider removing the DTS trigger, so LED is only used for USB "after boot".

Comment (mans0n) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Good point. Done.

**Comment by mans0n** [comment]
It seems like the `debounce-interval = <60>;` line has been originated from those old mach.c files in ar71xx and brcm63xx. In most mach files, the poll interval is set to 20ms and the debounce interval is set to three times the poll interval, which is 60ms.
I thought the longer debounce interval would help to reduce glitches, but after testing I found that the [default 5ms](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/input/gpio-keys.txt#L26) debounce interval is enough for most use cases.
So I decided to remove `debounce-interval` from my PRs.

**Comment by mans0n** [comment]
Rebased due to 19724e2.
What is blocking the merge of this PR?

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network [hunk 7]
```diff
@@ -199,6 +199,11 @@ ramips_setup_macs()
 	netgear,r6850)
 		wan_mac=$(macaddr_add "$(cat /sys/class/net/eth0/address)" 2)
 		;;
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
+		wan_mac=$(mtd_get_mac_binary factory 0x2e)
+		label_mac=$wan_mac
```
Comment (adrianschmutzler) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network]: I have just added a patch to master that sets lan+1 in an explicit case:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/board.d/02_network#L167

So, please rebase your patch onto that and _remove the wevo entries from this new node_.

Comment (adrianschmutzler) [reply target/linux/ramips/mt7621/base-files/etc/board.d/02_network]: I will merge this patch (patch 2) and rebase myself.

**Comment by mans0n** [comment]
Rebased remaining commits.

**Comment by adrianschmutzler** [comment]
Also update your initial comment here, maybe use strikethrough so the old stuff isn't just deleted.

**Comment by mans0n** [comment]
Thanks for your advice.

**Comment by mans0n** [comment]
I'm getting more and more negative about using USB LED for status indication (partly because I don't really use this device), so I'll withdraw the LED patch.
@adrianschmutzler Could you please merge the last remaining patch here? Actually that was the most essential part of this PR but it's been lingering too long.

**Comment by adrianschmutzler** [comment]
Merged.
